### PR TITLE
[JBPM-9411] Provide a way to access task details in boundary event

### PR DESCRIPTION
**Pull request by elguardian**
Jira: https://issues.redhat.com/browse/JBPM-9411

Add dataOuputs to boundary events.

### CODE REVIEW

**Comment by elguardian** [comment]
Jenkins retest  this please.

**Comment by elguardian** [comment]
Jenkins retest this please.

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            ksession.getWorkItemManager().completeWorkItem(workItemId, results);
+        } catch (Exception e) {
+            logger.warn("Work item could not be completed!");
+            e.printStackTrace();
+            fail(e.getClass().getSimpleName() + " thrown when completing work item: " + e.getMessage());
+        }
+    }
+
+    private void sleepAndVerifyTimerRuns(int processState) throws InterruptedException { 
+        // wait 3 seconds to see if the boss is notified
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java:124]: we're actually waiting for 2 seconds

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        KieSession ksession = runtimeEngine.getKieSession();
+        long ksessionId = ksession.getIdentifier();
+        assertTrue("session id not saved.", ksessionId > 0);
+        
+        HumanTaskMockHandler humanTaskMockHandler = new HumanTaskMockHandler();
+        ProcessInstance process = registerHTHandlerAndStartProcess(ksession, humanTaskMockHandler);
+        
+        sleepAndVerifyTimerRuns(process.getState());
+        List<? extends VariableInstanceLog> logs = getLogService().findVariableInstances(process.getId());
+        Assert.assertEquals(3, logs.size());
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java:82]: as per interval in timer is set to `500ms###1s` and we wait for 2 seconds, expected is `4` instead of `3`.
Though I would try to increase interval period, to avoid having any potential flaky test based on timing, just in case, wdyt?

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            logger.debug("Sleeping {} seconds", sleep / 1000);
+            Thread.sleep(sleep);
+            logger.debug("Awake!");
+        }
+       
+    }
+
+    
+
+    private static class HumanTaskMockHandler implements WorkItemHandler {
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java:136]: This class is registered as a work item handler, but never used afterwards in the test. Shouldn't we try to recover HumanTaskNodeInstance object from here as per requirements in BAPL? wdyt?

**Review by afalhambra — CHANGES_REQUESTED** [review]
Overall looks good to me, though I have some comments and questions

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+import org.kie.api.runtime.manager.RuntimeEngine;
+import org.kie.api.runtime.manager.audit.VariableInstanceLog;
+import org.kie.api.runtime.process.ProcessInstance;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import junit.framework.Assert;
+
+/**
+ */
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+public class BoundaryEventWithOutputSetPersistenceTest extends JbpmTestCase {
+
+    // General setup
+    private static final Logger logger = LoggerFactory.getLogger(BoundaryEventWithOutputSetPersistenceTest.class);
+
+    
+    private final static String PROCESS_FILE_NAME = "org/jbpm/test/functional/timer/BoundaryTimerProcessWithOutputSet.bpmn2";
+    private final static String PROCESS_NAME = "BoundaryTimerEventProcess";
+    private final static String WORK_ITEM_HANLDER_TASK = "Human Task";
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: done

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+
+    public BoundaryEventWithOutputSetPersistenceTest() { 
+        super(true, true);
+    }
+    
+    @Before
+    public void setup() { 
+        System.clearProperty(TIMER_FIRED_PROP);
+        System.clearProperty(TIMER_FIRED_TIME_PROP);
+    }
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: removed. don't needed anymore

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+        // The Human Task is completed
+        Map<String, Object> results = new HashMap<String, Object>();
+        try {
+            ksession.getWorkItemManager().completeWorkItem(workItemId, results);
+        } catch (Exception e) {
+            logger.warn("Work item could not be completed!");
+            e.printStackTrace();
+            fail(e.getClass().getSimpleName() + " thrown when completing work item: " + e.getMessage());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: corrected. IT is used.

File: jbpm-flow-builder/src/main/java/org/jbpm/process/builder/WorkItemNodeBuilder.java [hunk 8]
```diff
@@ -19,19 +19,26 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-flow-builder/src/main/java/org/jbpm/process/builder/WorkItemNodeBuilder.java:24]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java (hunk 2 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: it does not start in 0. but 500. anyway iwe are not testing that anymore. removed.

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java (hunk 1 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java (hunk 3 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: it is used now

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/timer/BoundaryEventWithOutputSetPersistenceTest.java]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 9]
```diff
@@ -0,0 +1,415 @@
...
+        long processId = process.getId();
+        assertTrue("process id not saved", processId > 0);
+
+        ksession.abortProcessInstance(processId);
+        // The process completes
+        process = ksession.getProcessInstance(process.getId());
+        assertNull("Expected process to have been completed and removed", process);
+        assertTrue(dataTaskMockHandler.getInputParameter("nodeInstance") instanceof HumanTaskNodeInstance);
+        assertTrue(dataTaskMockHandler.getInputParameter("event") instanceof java.lang.RuntimeException);
+        assertEquals(dataTaskMockHandler.getInputParameter("signal"), "Error-_2-401");
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:217]: Shouldn't we check for `workItem` as well?

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 10]
```diff
@@ -0,0 +1,415 @@
...
+    }
+    
+    
+    @Test
+    public void testBoundaryEscalationSignalAndCompleteHumanTask() throws InterruptedException {
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:222]: same as above

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryEscalationProcessWithOutputSet.bpmn2 [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+  <itemDefinition id="item_signal_ref_process" structureRef="String"  />
+  <itemDefinition id="item_event_ref_process" structureRef="Object"  />
+  <itemDefinition id="item_message_ref" structureRef="Object"  />
+  <escalation id="escalationRef" escalationCode="201" />
+	
+  <process processType="Private" 
+           isExecutable="true"
+           id="BoundaryEscalationEventProcess" 
+           name="Boundary Escalation Event Example" 
+           tns:packageName="org.jbpm.persistence.timer" >
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryEscalationProcessWithOutputSet.bpmn2:25]:
```suggestion
           tns:packageName="org.jbpm.persistence.escalation" >
```

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryEscalationProcessWithOutputSet.bpmn2 [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+			<targetRef>signalRef</targetRef>
+		</dataInputAssociation>
+	    <dataInputAssociation>
+			<sourceRef>eventProcess</sourceRef>
+			<targetRef>eventRef</targetRef>
+		</dataInputAssociation>
+		<dataInputAssociation>
+			<sourceRef>workItemProcess</sourceRef>
+			<targetRef>workItemRef</targetRef>
+		</dataInputAssociation>
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryEscalationProcessWithOutputSet.bpmn2:107]: `workItemProcess` is not defined in this process

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+        
+        createRuntimeManager(PROCESS_ESCALATION_FILE_NAME);
+
+        RuntimeEngine runtimeEngine = getRuntimeEngine();
+        KieSession ksession = runtimeEngine.getKieSession();
+        long ksessionId = ksession.getIdentifier();
+        assertTrue("session id not saved.", ksessionId > 0);
+        
+        HumanTaskMockHandler humanTaskMockHandler = new HumanTaskMockHandler(true);
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:231]: This is not needed for this test

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+        assertEquals("Expected process state to be " + processStateName[ProcessInstance.STATE_COMPLETED] + " not "
+                + processStateName[processState], ProcessInstance.STATE_COMPLETED, processState);
+
+        
+        // The process completes
+        process = ksession.getProcessInstance(process.getId());
+        assertNull("Expected process to have been completed and removed", process);
+        assertTrue(dataTaskMockHandler.getInputParameter("nodeInstance") instanceof CompositeContextNodeInstance);
+        assertEquals(dataTaskMockHandler.getInputParameter("signal"), "Escalation-_2-201");
+        assertNull(dataTaskMockHandler.getInputParameter("event"));
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:252]: Shouldn't we check `workItem` as well?

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryCompensationProcessWithOutputSet.bpmn2 [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+             
+  <itemDefinition id="item_nodeInstance_ref_process" structureRef="String"  />
+  <itemDefinition id="item_signal_ref_process" structureRef="String"  />
+  <itemDefinition id="item_event_ref_process" structureRef="Object"  />
+	
+  <process processType="Private" 
+           isExecutable="true"
+           id="BoundaryCompensationEventProcess" 
+           name="Boundary Compensation Event Example" 
+           tns:packageName="org.jbpm.persistence.timer" >
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryCompensationProcessWithOutputSet.bpmn2:23]:
```suggestion
           tns:packageName="org.jbpm.persistence.compensation" >
```

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryCompensationProcessWithOutputSet.bpmn2 [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+		<dataInputAssociation>
+			<sourceRef>signalProcess</sourceRef>
+			<targetRef>signalRef</targetRef>
+		</dataInputAssociation>
+	    <dataInputAssociation>
+			<sourceRef>eventProcess</sourceRef>
+			<targetRef>eventRef</targetRef>
+		</dataInputAssociation>
+		<dataInputAssociation>
+			<sourceRef>workItemProcess</sourceRef>
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryCompensationProcessWithOutputSet.bpmn2:103]: `workItemProcess` is not defined in this process

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 17]
```diff
@@ -0,0 +1,415 @@
...
+    }
+    
+    @Test
+    public void testBoundaryCompensationSignalAndCompleteHumanTask() throws InterruptedException {
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:256]: same as above

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+        ksession.signalEvent("Compensation", "_2");
+        
+        
+        completeWork(ksession, humanTaskMockHandler);
+        // The process completes
+        process = ksession.getProcessInstance(process.getId());
+        assertNull("Expected process to have been completed and removed", process);
+        assertNull(dataTaskMockHandler.getInputParameter("nodeInstance"));
+        assertEquals(dataTaskMockHandler.getInputParameter("event"), "_2");
+        assertEquals(dataTaskMockHandler.getInputParameter("signal"), "Compensation");
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:290]: shouldn't we check `workItem` as well?

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+    }
+    
+    @Test
+    public void testBoundaryConditionalSignalAndCompleteHumanTask() throws InterruptedException {
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:294]: same as above

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2 [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+  <itemDefinition id="item_signal_ref_process" structureRef="String"  />
+  <itemDefinition id="item_event_ref_process" structureRef="Object"  />
+  <itemDefinition id="item_message_ref" structureRef="Object"  />
+  <escalation id="escalationRef" escalationCode="201" />
+	
+  <process processType="Private" 
+           isExecutable="true"
+           id="BoundaryConditionalEventProcess" 
+           name="Boundary Conditional Event Example" 
+           tns:packageName="org.jbpm.persistence.timer" >
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2:25]:
```suggestion
           tns:packageName="org.jbpm.persistence.conditional" >
```

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2 [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+             xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI"
+             xmlns:dc="http://www.omg.org/spec/DD/20100524/DC"
+             xmlns:di="http://www.omg.org/spec/DD/20100524/DI"
+             xmlns:tns="http://www.jboss.org/drools">
+             
+  <itemDefinition id="item_nodeInstance_ref_process" structureRef="String"  />
+  <itemDefinition id="item_signal_ref_process" structureRef="String"  />
+  <itemDefinition id="item_event_ref_process" structureRef="Object"  />
+  <itemDefinition id="item_message_ref" structureRef="Object"  />
+  <escalation id="escalationRef" escalationCode="201" />
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2:19]: Not needed

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2 [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+          <outputSet></outputSet>
+        </ioSpecification>
+      </userTask>
+      
+  
+    <endEvent id="_3" name="End Main Process">
+      <terminateEventDefinition/>
+    </endEvent>
+  
+    <boundaryEvent id="_6" name="Timer Event" attachedToRef="_2" cancelActivity="false">
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2:46]:
```suggestion
    <boundaryEvent id="_6" name="Conditional Event" attachedToRef="_2" cancelActivity="false">
```

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2 [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+			<targetRef>signalRef</targetRef>
+		</dataInputAssociation>
+	    <dataInputAssociation>
+			<sourceRef>eventProcess</sourceRef>
+			<targetRef>eventRef</targetRef>
+		</dataInputAssociation>
+		<dataInputAssociation>
+			<sourceRef>workItemProcess</sourceRef>
+			<targetRef>workItemRef</targetRef>
+		</dataInputAssociation>
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2:102]: `workItemProcess` is not defined in this process

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 24]
```diff
@@ -0,0 +1,415 @@
...
+        ksession.signalEvent("RuleFlowStateEvent-" + PROCESS_CONDITIONAL_NAME + "-4-_2" , new org.jbpm.test.entity.Person(1L, "john"));
+       
+        completeWork(ksession, humanTaskMockHandler);
+
+        // The process completes
+        process = ksession.getProcessInstance(process.getId());
+        assertNull("Expected process to have been completed and removed", process);
+        assertTrue(dataTaskMockHandler.getInputParameter("nodeInstance") instanceof HumanTaskNodeInstance);
+        assertTrue(dataTaskMockHandler.getInputParameter("event") instanceof Person);
+        assertEquals(dataTaskMockHandler.getInputParameter("signal"), "RuleFlowStateEvent-BoundaryConditionalEventProcess-4-_2");
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:327]: Shouldn't we check `workItem` here as well?

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java [hunk 25]
```diff
@@ -0,0 +1,415 @@
...
+        }
+
+    }
+
+
+
+    private static class HumanTaskMockHandler implements WorkItemHandler {
+
+        private org.kie.api.runtime.process.WorkItemManager workItemManager;
+        private org.kie.api.runtime.process.WorkItem workItem;
```
Comment (afalhambra) [inline jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:364]: we don't need to use fully qualified class name for them

**Review by afalhambra — COMMENTED** [review]
Few minor comments, but it is a very good job

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+
+import org.jbpm.workflow.core.node.Assignment;
+import org.jbpm.workflow.core.node.DataAssociation;
+import org.jbpm.workflow.core.node.Transformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class DataAssociationFactory {
+    
+    protected static final Logger logger = LoggerFactory.getLogger(DataAssociationFactory.class);
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java:31]: done

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+    
+    private DataAssociationFactory() {
+        // do nothing
+    }
+
+    public static DataAssociation readDataOutputAssociation(org.w3c.dom.Node xmlNode, Map<String, String> dataOutputs) {
+
+        org.w3c.dom.NodeList nodeList = xmlNode.getChildNodes();
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java:39]: none of them can be null as per design of node handler

[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java:39]: removed.

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java [hunk 28]
```diff
@@ -0,0 +1,98 @@
...
+                    assignment.add(readAssignment(subNode));
+                    break;
+                }
+            }
+
+        }
+        return new DataAssociation(source, target, assignment, transformation);
+    }
+    
+    public static Assignment readAssignment(org.w3c.dom.Node xmlNode) {
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java:80]: it is ok. as assignment can only be within data associations.

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java [hunk 29]
```diff
@@ -0,0 +1,98 @@
...
+                    String expression = subNode.getTextContent();
+                    transformation = new Transformation(lang, expression, source);
+                    break;
+                }
+                case "assignment":
+                {
+                    assignment.add(readAssignment(subNode));
+                    break;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (elguardian) [reply jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/elements/DataAssociationFactory.java:74]: it is not possible to reach another state as it only can be from or to. Parser is throwing already a warning or error depending of the situation.

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2 (hunk 22 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2:46]: done

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2 (hunk 23 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2:102]: added.

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java (hunk 25 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:364]: removed

File: jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java (hunk 24 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/java/org/jbpm/test/functional/event/BoundaryEventWithOutputSetPersistenceTest.java:327]: done

File: jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2 (hunk 21 above)
Comment (elguardian) [reply jbpm-test-coverage/src/test/resources/org/jbpm/test/functional/event/BoundaryConditionalProcessWithOutputSet.bpmn2:20]: removed
